### PR TITLE
Enable P2P memory copy

### PR DESCRIPTION
**Pull request by panyx0718**
On k40 with 4 devices, time reduces from ~4.0 to ~3.8+, should be
more obvious on better hardware

### CODE REVIEW

File: paddle/fluid/framework/init.cc [hunk 1]
```diff
@@ -62,6 +62,18 @@ void InitDevices() {
 
   for (int i = 0; i < count; ++i) {
     places.emplace_back(platform::CUDAPlace(i));
+    for (int j = 0; j < count; ++j) {
+      if (i == j) continue;
+      int can_acess = -1;
+      PADDLE_ENFORCE(cudaDeviceCanAccessPeer(&can_acess, i, j),
+                     "Failed to test P2P access.");
+      if (can_acess != 1) {
+        VLOG(1) << "Cannot enable P2P access from " << i << " to " << j;
```
Comment (chengduoZH) [inline paddle/fluid/framework/init.cc]: It may be better to use `LOG(WARNING)`.

File: paddle/fluid/framework/init.cc [hunk 2]
```diff
@@ -62,6 +62,18 @@ void InitDevices() {
...
+      } else {
+        cudaSetDevice(i);
+        cudaDeviceEnablePeerAccess(j, 0);
```
Comment (dzhwinter) [inline paddle/fluid/framework/init.cc]: Is P2P Access need any specific hardware? I vaguely remember that it needs the slim-connect(or something)?

File: paddle/fluid/framework/init.cc (hunk 1 above)
Comment (dzhwinter) [reply paddle/fluid/framework/init.cc]: Agreed. VLOG for debugging only.  https://github.com/PaddlePaddle/Paddle/pull/5181
LOG(WARNING) is better here.

Comment (panyx0718) [reply paddle/fluid/framework/init.cc]: Done

File: paddle/fluid/framework/init.cc (hunk 2 above)
Comment (panyx0718) [reply paddle/fluid/framework/init.cc]: It needs to be tesla. So gforce titan x doesn't support it

**Pull request merged**